### PR TITLE
[tools] Fix silently fail when autodep fails

### PR DESCRIPTION
**Pull request by KB5201314**
This commit fixes the bug we mentioned here: https://github.com/occlum/occlum/pull/975#discussion_r942031420

### CODE REVIEW

**Comment by volcano0dr** [comment]
@KB5201314 Thanks for your contribution.
CI test result shows `Local_attestaion` test failed, can you help to fix this?
Error message: 
`[2022-08-15T00:56:02Z ERROR copy_bom::util] cannot autodep for ../AppResponder/responder. exit status: 127 stderr: Error loading shared library libecdh.so: No such file or directory (needed by ../AppResponder/responder)`

**Comment by guzongmin** [comment]
Maybe the BOM file needs an update. This patch exposes the BOM file issue.

**Comment by qzheng527** [comment]
@KB5201314 @guzongmin @volcano0dr with this new patch, some previous working demos will fail on copy_bom stage,  including all the demos/example which clearly specify the library path on BOM file but not setting correct LD_LIBRARY_PATH, such as [local_attestation](https://github.com/occlum/occlum/blob/master/demos/local_attestation/bom.yaml#L14), and [init_ra_flow](https://github.com/occlum/occlum/blob/master/demos/remote_attestation/init_ra_flow/init_ra_client.yaml#L11).
How about just throwing some clear warning messages instead of error exit?
Just let the users to decide which way is better? Either setting correct LD_LIBRARY_PATH or clearly specify the library path on BOM file.

File: tools/copy_bom/src/error.rs [hunk 1]
```diff
@@ -6,3 +6,4 @@ pub static CREATE_DIR_ERROR: i32 = -4;
 pub static CREATE_SYMLINK_ERROR: i32 = -5;
 pub static COPY_DIR_ERROR: i32 = -6;
 pub static INCORRECT_HASH_ERROR: i32 = -7;
+pub static AUTODEP_ERROR: i32 = -7;
```
Comment (qzheng527) [inline tools/copy_bom/src/error.rs:9]: @KB5201314 Please use different error value (-7).

**Comment by guzongmin** [comment]
> @KB5201314 @guzongmin @volcano0dr with this new patch, some previous working demos will fail on copy_bom stage, including all the demos/example which clearly specify the library path on BOM file but not setting correct LD_LIBRARY_PATH, such as [local_attestation](https://github.com/occlum/occlum/blob/master/demos/local_attestation/bom.yaml#L14), and [init_ra_flow](https://github.com/occlum/occlum/blob/master/demos/remote_attestation/init_ra_flow/init_ra_client.yaml#L11). How about just throwing some clear warning messages instead of error exit? Just let the users to decide which way is better? Either setting correct LD_LIBRARY_PATH or clearly specify the library path on BOM file.

Thanks. I get the reason now. So the situation is the same as the previous one @KB5201314 PR wants to address. In the local attestation demo, the "responder" depends on the "libecdh.so". The user just adds libecdh.so into the bom file instead of setting the LD_LIBRARY_PATH.

I think using LD_LIBRARY_PATH is better than manually adding the .so file into the bom file. Since there are only two demos has such an issue, can we simply add the LD_LIBRARY_PATH when using the copy_bom command?

**Comment by qzheng527** [comment]
> > @KB5201314 @guzongmin @volcano0dr with this new patch, some previous working demos will fail on copy_bom stage, including all the demos/example which clearly specify the library path on BOM file but not setting correct LD_LIBRARY_PATH, such as [local_attestation](https://github.com/occlum/occlum/blob/master/demos/local_attestation/bom.yaml#L14), and [init_ra_flow](https://github.com/occlum/occlum/blob/master/demos/remote_attestation/init_ra_flow/init_ra_client.yaml#L11). How about just throwing some clear warning messages instead of error exit? Just let the users to decide which way is better? Either setting correct LD_LIBRARY_PATH or clearly specify the library path on BOM file.
> 
> Thanks. I get the reason now. So the situation is the same as the previous one @KB5201314 PR wants to address. In the local attestation demo, the "responder" depends on the "libecdh.so". The user just adds libecdh.so into the bom file instead of setting the LD_LIBRARY_PATH.
> 
> I think using LD_LIBRARY_PATH is better than manually adding the .so file into the bom file. Since there are only two demos has such an issue, can we simply add the LD_LIBRARY_PATH when using the copy_bom command?

Updating the two demos is simple, but I have two concerns.

1. Perhaps some users already used the "adding the .so file into the bom file" way. 

2. LD_LIBRARY_PATH is soemtimes too powerful. Its prioriy precedes the directories defined in /etc/ld.so.conf, and also /lib, /usr/lib. What if there are same name libraries in LD_LIBRARY_PATH but users actually expects the libraries under /lib ?

**Comment by guzongmin** [comment]
> > > @KB5201314 @guzongmin @volcano0dr with this new patch, some previous working demos will fail on copy_bom stage, including all the demos/example which clearly specify the library path on BOM file but not setting correct LD_LIBRARY_PATH, such as [local_attestation](https://github.com/occlum/occlum/blob/master/demos/local_attestation/bom.yaml#L14), and [init_ra_flow](https://github.com/occlum/occlum/blob/master/demos/remote_attestation/init_ra_flow/init_ra_client.yaml#L11). How about just throwing some clear warning messages instead of error exit? Just let the users to decide which way is better? Either setting correct LD_LIBRARY_PATH or clearly specify the library path on BOM file.
> > 
> > 
> > Thanks. I get the reason now. So the situation is the same as the previous one @KB5201314 PR wants to address. In the local attestation demo, the "responder" depends on the "libecdh.so". The user just adds libecdh.so into the bom file instead of setting the LD_LIBRARY_PATH.
> > I think using LD_LIBRARY_PATH is better than manually adding the .so file into the bom file. Since there are only two demos has such an issue, can we simply add the LD_LIBRARY_PATH when using the copy_bom command?
> 
> Updating the two demos is simple, but I have two concerns.
> 
> 1. Perhaps some users already used the "adding the .so file into the bom file" way.
> 2. LD_LIBRARY_PATH is soemtimes too powerful. Its prioriy precedes the directories defined in /etc/ld.so.conf, and also /lib, /usr/lib. What if there are same name libraries in LD_LIBRARY_PATH but users actually expects the libraries under /lib ?

I think there are some ways to address the issues:
1. Perhaps some users already used the "adding the .so file into the bom file" way.
We can put the additional message "Error loading shared library libpython3.7m.so.1.0: No such file or directory (needed by /root/alpine_python/usr/local/bin/python3.7). \n It seems that some required shared object is missed in the LD_LIBRARY_PATH. Please copy the file into one of the folders in the LD_LIBRARY_PATH or add the shared object folder into the  LD_LIBRARY_PATH."

2. LD_LIBRARY_PATH is soemtimes too powerful. 
I am not suggesting replacing the LD_LIBRARY_PATH with a new one. I suggest using "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:<one additional folder>" to add one more folder to help the copy_bom find the missed file.

**Comment by qzheng527** [comment]
> 2\. LD_LIBRARY_PATH is soemtimes too powerful.
>     I am not suggesting replacing the LD_LIBRARY_PATH with a new one. I suggest using "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:<one additional folder>" to add one more folder to help the copy_bom find the missed file.

I mean the LD_LIBRARY_PATH precedes the default search path defined in /etc/ld.so.conf or  /lib and /usr/lib which are not defined by LD_LIBRARY_PATH.

**Comment by guzongmin** [comment]
> > 2. LD_LIBRARY_PATH is soemtimes too powerful.
> > I am not suggesting replacing the LD_LIBRARY_PATH with a new one. I suggest using "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:" to add one more folder to help the copy_bom find the missed file.
> 
> I mean the LD_LIBRARY_PATH precedes the default search path defined in /etc/ld.so.conf or /lib and /usr/lib which are not defined by LD_LIBRARY_PATH.



> > 2. LD_LIBRARY_PATH is soemtimes too powerful.
> > I am not suggesting replacing the LD_LIBRARY_PATH with a new one. I suggest using "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:" to add one more folder to help the copy_bom find the missed file.
> 
> I mean the LD_LIBRARY_PATH precedes the default search path defined in /etc/ld.so.conf or /lib and /usr/lib which are not defined by LD_LIBRARY_PATH.

I see. Can we do it like:
LD_LIBRARY_PATH=$LD_LIBRARY_PATH:\<one additional folder\> copy_bom xxx 
then it would not impact other apps

**Comment by qzheng527** [comment]
> > > 2. LD_LIBRARY_PATH is soemtimes too powerful.
> > >    I am not suggesting replacing the LD_LIBRARY_PATH with a new one. I suggest using "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:" to add one more folder to help the copy_bom find the missed file.
> > 
> > 
> > I mean the LD_LIBRARY_PATH precedes the default search path defined in /etc/ld.so.conf or /lib and /usr/lib which are not defined by LD_LIBRARY_PATH.
> 
> > > 2. LD_LIBRARY_PATH is soemtimes too powerful.
> > >    I am not suggesting replacing the LD_LIBRARY_PATH with a new one. I suggest using "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:" to add one more folder to help the copy_bom find the missed file.
> > 
> > 
> > I mean the LD_LIBRARY_PATH precedes the default search path defined in /etc/ld.so.conf or /lib and /usr/lib which are not defined by LD_LIBRARY_PATH.
> 
> I see. Can we do it like: LD_LIBRARY_PATH=$LD_LIBRARY_PATH:<one additional folder> copy_bom xxx then it would not impact other apps

Yes, this way can work for most of the case.
But the LD_LIBRARY_PATH still take effect on the lib/bin defines in the bom file.
For example,  there is a bin file defined in a bom file.
This bin file expects some libs (such as libcrypto.so.1.1) in /lib and also one libexample.so in /home/admin/local/lib/.
But what if there is one libcrypto.so.1.1 in /home/admin/local/lib/ as well?
In this case, LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/admin/local/lib/ copy_bom xxx will include the /home/admin/local/lib/libcrypto.so.1.1 instead of the /lib/libcrypto.so.1.1.
I met such kind of issue when using some libs by extracting them from the docker image rootfs.



the bin expects the libs in /lib and also the libexample.so in

**Comment by guzongmin** [comment]
> > > > 2. LD_LIBRARY_PATH is soemtimes too powerful.
> > > >    I am not suggesting replacing the LD_LIBRARY_PATH with a new one. I suggest using "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:" to add one more folder to help the copy_bom find the missed file.
> > > 
> > > 
> > > I mean the LD_LIBRARY_PATH precedes the default search path defined in /etc/ld.so.conf or /lib and /usr/lib which are not defined by LD_LIBRARY_PATH.
> > 
> > 
> > > > 2. LD_LIBRARY_PATH is soemtimes too powerful.
> > > >    I am not suggesting replacing the LD_LIBRARY_PATH with a new one. I suggest using "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:" to add one more folder to help the copy_bom find the missed file.
> > > 
> > > 
> > > I mean the LD_LIBRARY_PATH precedes the default search path defined in /etc/ld.so.conf or /lib and /usr/lib which are not defined by LD_LIBRARY_PATH.
> > 
> > 
> > I see. Can we do it like: LD_LIBRARY_PATH=$LD_LIBRARY_PATH: copy_bom xxx then it would not impact other apps
> 
> Yes, this way can work for most of the case. But the LD_LIBRARY_PATH still take effect on the lib/bin defines in the bom file. For example, there is a bin file defined in a bom file. This bin file expects some libs (such as libcrypto.so.1.1) in /lib and also one libexample.so in /home/admin/local/lib/. But what if there is one libcrypto.so.1.1 in /home/admin/local/lib/ as well? In this case, LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/admin/local/lib/ copy_bom xxx will include the /home/admin/local/lib/libcrypto.so.1.1 instead of the /lib/libcrypto.so.1.1. I met such kind of issue when using some libs by extracting them from the docker image rootfs.

First, I want to explain why I insist on returning an error. In the current Occlum instance building process, we do not give the user enough support to build a correct instance. For those experienced users, it is not a problem, they could read the log file and find the missing files. But for most users, it is a hard task. 
Then, let us find a solution. Actually, the copy_bom tool has another feature (createlinks) to create a soft link in the image. So the user could use the prefix LD_LIBRARY_PATH to add the additional shared objects into the images and then add create a soft link to help the loader inside Occlum to find the shared object.

**Comment by KB5201314** [comment]
Sorry to reply so late.

> I think using LD_LIBRARY_PATH is better than manually adding the .so file into the bom file. Since there are only two demos has such an issue, can we simply add the LD_LIBRARY_PATH when using the copy_bom command?

Things like `LD_LIBRARY_PATH=$LD_LIBRARY_PATH:<one additional folder> copy_bom` won't work, Because copy_bom will overwrite this env variable for `ld.so`. To make it work, we also need to make a change to this code snippet:
https://github.com/occlum/occlum/blob/5c10af738e378c8703e757832b62639befd0107d/tools/copy_bom/src/util.rs#L260-L272 


> But the LD_LIBRARY_PATH still take effect on the lib/bin defines in the bom file.
For example, there is a bin file defined in a bom file.
This bin file expects some libs (such as libcrypto.so.1.1) in /lib and also one libexample.so in /home/admin/local/lib/.
But what if there is one libcrypto.so.1.1 in /home/admin/local/lib/ as well?
In this case, LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/admin/local/lib/ copy_bom xxx will include the /home/admin/local/lib/libcrypto.so.1.1 instead of the /lib/libcrypto.so.1.1.
I met such kind of issue when using some libs by extracting them from the docker image rootfs.

I think a simple solution to this problem is to split the bom.yaml file into multiple parts and run copy_bom multiple times.

> Then, let us find a solution. Actually, the copy_bom tool has another feature (createlinks) to create a soft link in the image. So the user could use the prefix LD_LIBRARY_PATH to add the additional shared objects into the images and then add create a soft link to help the loader inside Occlum to find the shared object.

Do you mean that the user needs to add a symlink to the additional shared objects file in bom.yaml?

**Comment by guzongmin** [comment]
We do need to discuss the design goal of the copy_bom tool. https://github.com/occlum/occlum/issues/565 is the original design

What I believe are: 
1. The BOM file should not be split. Because it is the best way to help users maintain the content of the Occlum.
2. The tool should report all the missed dependencies. Because it is the best place to report those issues.

**Comment by guzongmin** [comment]
"For example, there is a bin file defined in a bom file.
This bin file expects some libs (such as libcrypto.so.1.1) in /lib and also one libexample.so in /home/admin/local/lib/.
But what if there is one libcrypto.so.1.1 in /home/admin/local/lib/ as well?
In this case, LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/admin/local/lib/ copy_bom xxx will include the /home/admin/local/lib/libcrypto.so.1.1 instead of the /lib/libcrypto.so.1.1."

For this case, my opinion is that: The assumption "what if there is one libcrypto.so.1.1 in /home/admin/local/lib/ as well" is very strange. If the libcrypto.so.1.1 in /home/admin/local/lib/ is not a workable shared object, why would the user put it in the PATH? And we can find another case: the libcrypto.so.1.1 in /lib/ is not a valid one, and the user provided in  /home/admin/local/lib/ is the lib user want to use inside in Occlum. In order to solve such an issue, the copy_bom tool should print detail information about the source path.

**Comment by qzheng527** [comment]
> If the libcrypto.so.1.1 in /home/admin/local/lib/ is not a workable shared object, why would the user put it in the PATH?
For example, the user get most of the libs required but missed several libs whose building may take very long time.
So the user pull a docker image (included the missed libs) and unpack it in some place such as /home/admin/rootfs/.
The missed libs are in the /home/admin/rootfs/lib but there is one unexpected libcrypto.so.1.1 in /home/admin/rootfs/lib as well.

The case is not important, I think the key point here is do we need give below option to the users?
"The copy_bom can work (warning messages if any, let the user do the judgement) by clearly specifying the libs in the bom file but not setting all the library path by LD_LIBRARY_PATH."
 My opinion is it should work otherwise why we call the tool copy_**bom**?

**Comment by KB5201314** [comment]
> > If the libcrypto.so.1.1 in /home/admin/local/lib/ is not a workable shared object, why would the user put it in the PATH?
> > For example, the user get most of the libs required but missed several libs whose building may take very long time.
> > So the user pull a docker image (included the missed libs) and unpack it in some place such as /home/admin/rootfs/.
> > The missed libs are in the /home/admin/rootfs/lib but there is one unexpected libcrypto.so.1.1 in /home/admin/rootfs/lib as well.
> 
> The case is not important, I think the key point here is do we need give below option to the users? "The copy_bom can work (warning messages if any, let the user do the judgement) by clearly specifying the libs in the bom file but not setting all the library path by LD_LIBRARY_PATH." My opinion is it should work otherwise why we call the tool copy_**bom**?

I like the idea of having the bom.yaml file describe everything, including describing a missing lib in bom.yaml. Actually, the test that failed did the same thing (It added the missing libecdh.so to bom.yaml).
https://github.com/occlum/occlum/blob/5c10af738e378c8703e757832b62639befd0107d/demos/local_attestation/bom.yaml#L1-L14

Maybe we should add a field to the bom file to allow autodep to copy the discovered libs, but leave the missing libs to the user.

**Comment by StevenJiang1110** [comment]
Sorry to reply so late. 
It is **very difficult** for `copy_bom` to find all dependencies by setting LD_LIBRARY_PATH. For example, for projects that depend on JDK, we only need to copy a JDK folder at present. But in fact, the libraries that executable files in JDK directories depend on may exist in multiple directories in the JDK folder. If you want to find all dependent libraries through LD_LIBRARY_PATH, you must add all the paths containing these libraries to LD_LIBRARY_PATH, which will be a long list. 

Of course this issue reveals existing problems in copy_bom. When copy_bom cannot find the library, it will generate warning information, but it will only be seen when the log is opened. For the case where the log is not opened, these warnings are directly skipped. This is something that can be improved.

However, before opening the warning, it is necessary to filter the warning information. For the case where some libraries are not found, copy_bom needs to determine whether these libraries are explicitly declared in the bom file, so the libraries in warning message is indeed missed, but this It is not easy to implement, because copy_bom does not record which libraries are copied when copying a folder, and there can be libraries with same name but different content.

File: tools/copy_bom/src/error.rs (hunk 1 above)
Comment (KB5201314) [reply tools/copy_bom/src/error.rs]: Fixed now :)

**Review by StevenJiang1110 — COMMENTED** [review]
Thanks for your contribution. I think it makes sense to abort copy_bom when autodep for  **copyfiles** fails (as the PR indicates, ***NOT*** including files in copydirs). However, before aborting copy_bom, we should check whether missing libraries are manually added in bom file. By this way, this PR will not affect current demos.

**Comment by guzongmin** [comment]
> Thanks for your contribution. I think it makes sense to abort copy_bom when autodep for **copyfiles** fails (as the PR indicates, _**NOT**_ including files in copydirs). However, before aborting copy_bom, we should check whether missing libraries are manually added in bom file. By this way, this PR will not affect current demos.

@StevenJiang1110 Thanks for your comments. It seems a good solution. Do you know what is the complexity to add the final check before returning the error?

**Comment by guzongmin** [comment]
> Thanks for your contribution. I think it makes sense to abort copy_bom when autodep for **copyfiles** fails (as the PR indicates, _**NOT**_ including files in copydirs). However, before aborting copy_bom, we should check whether missing libraries are manually added in bom file. By this way, this PR will not affect current demos.

@StevenJiang1110 Thanks for your comments. It seems a good solution. Do you know what is the complexity to add the final check before returning the error?

**Comment by StevenJiang1110** [comment]
> @StevenJiang1110 Thanks for your comments. It seems a good solution. Do you know what is the complexity to add the final check before returning the error?

I think the most complicated case is that user may include these libraries in **copydirs**, but `copy_bom` itself won't check the content of **copydirs**. I think there may be two possible solutions. First is that `copy_bom` itself check all libraries in **copydirs**. Second is to require user to add these libraries in **copyfiles**, then `copy_bom` won't need to check each **copydir**. I'm not sure which choice is more user-friendly.

**Comment by guzongmin** [comment]
> > @StevenJiang1110 Thanks for your comments. It seems a good solution. Do you know what is the complexity to add the final check before returning the error?
> 
> I think the most complicated case is that user may include these libraries in **copydirs**, but `copy_bom` itself won't check the content of **copydirs**. I think there may be two possible solutions. First is that `copy_bom` itself check all libraries in **copydirs**. Second is to require user to add these libraries in **copyfiles**, then `copy_bom` won't need to check each **copydir**. I'm not sure which choice is more user-friendly.

I more like the second one. Users should use copy_bom to build the instance. It may need some time to learn. But it would help the user save a lot of debugging time to find the missing files. I think it is worth learning copy_bom.

**Comment by qzheng527** [comment]
How about we do this way?
Let copy_bom do all based on the bom file, when some libraries are not found, just record them but not quit, continue till complete processing the bom file. Then, check if the recorded missing libraries are in the final image. If yes, copy_bom pass. If not,  return error, and print what libraries are missing.

**Comment by KB5201314** [comment]
> How about we do this way? Let copy_bom do all based on the bom file, when some libraries are not found, just record them but not quit, continue till complete processing the bom file. Then, check if the recorded missing libraries are in the final image. If yes, copy_bom pass. If not, return error, and print what libraries are missing.

How do we check if the missing library is in the oclum image? By checking if the file with the corresponding filename exists in **copyfiles** list?

**Comment by qzheng527** [comment]
> > How about we do this way? Let copy_bom do all based on the bom file, when some libraries are not found, just record them but not quit, continue till complete processing the bom file. Then, check if the recorded missing libraries are in the final image. If yes, copy_bom pass. If not, return error, and print what libraries are missing.
> 
> How do we check if the missing library is in the oclum image? By checking if the file with the corresponding filename exists in **copyfiles** list?

All the generated files (if not quit) are in the final occlum image dir defined by `"--root"` which it is `occlum_instance/image` in default. If there are missing libraries recorded, the final step is to check if they are already in the `occlum_instance/image` (checking the filenames only). If all found, treat the copy_bom successful. If not, error return, print the missing libraries.

**Comment by guzongmin** [comment]
> > > How about we do this way? Let copy_bom do all based on the bom file, when some libraries are not found, just record them but not quit, continue till complete processing the bom file. Then, check if the recorded missing libraries are in the final image. If yes, copy_bom pass. If not, return error, and print what libraries are missing.
> > 
> > 
> > How do we check if the missing library is in the oclum image? By checking if the file with the corresponding filename exists in **copyfiles** list?
> 
> All the generated files (if not quit) are in the final occlum image dir defined by `"--root"` which it is `occlum_instance/image` in default. If there are missing libraries recorded, the final step is to check if they are already in the `occlum_instance/image` (checking the filenames only). If all found, treat the copy_bom successful. If not, error return, print the missing libraries.

I think copy_bom should not depend on the files in the image folder because the original folder maybe not be empty when using the copy_bom tool. The uncertainty would impact the reproducibility. I remember that the copy_bom tool would catch the files copied in a list. So I think we only need to check the missed files in the list before return failure.

**Comment by guzongmin** [comment]
I believe the copy_bom tool is essential because it links the user with Occlum. The BOM file is the interface, and it works well as expected. We need to make the tool better, easy to use, stable, and report precise error messages to help users adjust the BOM file. 

Thanks for @KB5201314 submitting this PR and the developers who care about the tool to discuss improving this tool.

We already discussed a lot in the PR. Maybe someone could write an RFC to discuss how to improve the copy_bom tool before we write the code.

**Comment by guzongmin** [comment]
Fixed in https://github.com/occlum/occlum/pull/1003
Close this PR.
@KB5201314 Thank you very much. I hope you can bring more PRs to help the Occlum project.

**Pull request closed without merging**